### ISSUE TITLE
Checking for a Julia installation before running install_julia?

### ISSUE THREAD

**Issue by ChrisRackauckas** [Non-Contradiction/JuliaCall#149]
In diffeqr I want to do:

```R
if !julia_is_installed()
  install_julia()
end
```

inside of `diffeqr::diffeq_setup()`. Is there a way to check for whether Julia is installed?

**Comment by Non-Contradiction** [comment]
Currently there is an internal function to locate the folder for julia binary called `JuliaCall:::julia_locate(JULIA_HOME = NULL)`.
The function will return  the folder for julia binary if julia is found, and will return `NULL` if julia is not found.
I will keep this internal function, and may export it in the future. So it could be used without worry.

Since it is currently an internal function only, to get around CRAN's check, you can use the following trick in the function you want to use it
```
julia_locate <- do.call(":::", list("JuliaCall", quote(julia_locate)))
```

**Comment by ChrisRackauckas** [comment]
Hmm, I'm getting some issues.

```julia
> de <- diffeqr::diffeq_setup()
Julia version 1.5.0 at location C:\Users\accou\AppData\Local\JULIAC~1\JULIAC~1\julia\V15~1.0\bin will be used.
Loading setup script for JuliaCall...
Finish loading setup script for JuliaCall.
 Error: Error happens in Julia.
InitError: could not load library "libdSFMT"
The specified module could not be found. 
Stacktrace:
 [1] dsfmt_init_by_array at D:\buildbot\worker\package_win64\build\usr\share\julia\stdlib\v1.5\Random\src\DSFMT.jl:70 [inlined]
 [2] seed!(::Random.MersenneTwister, ::Array{UInt32,1}) at D:\buildbot\worker\package_win64\build\usr\share\julia\stdlib\v1.5\Random\src\RNGs.jl:283
 [3] seed! at D:\buildbot\worker\package_win64\build\usr\share\julia\stdlib\v1.5\Random\src\RNGs.jl:290 [inlined]
 [4] seed! at D:\buildbot\worker\package_win64\build\usr\share\julia\stdlib\v1.5\Random\src\Random.jl:426 [inlined]
 [5] Random.MersenneTwister(::Nothing) at D:\buildbot\worker\package_win64\build\usr\share\julia\stdlib\v1.5\Random\src\RNGs.jl:139 (repeats 2 times)
 [6] default_rng(::Int64) at D:\buildbot\worker\package_win64\build\usr\share\julia\stdlib\v1.5\Random\src\RNGs.jl:304
 [7] default_rng at D:\buildbot\worker\package_win64\build\usr\share\julia\stdlib\v1.5\Random 
```

**Comment by ChrisRackauckas** [comment]
```julia
 Error: Error happens in Julia.
could not load library "libopenlibm"
The specified module could not be found. 
Stacktrace:
 [1] log10(::Float64) at .\math.jl:566
 [2] ode_determine_initdt(::Array{Float64,1}, ::Float64, ::Float64, ::Float64, ::Float64, ::Float64, ::typeof(DiffEqBase.ODE_DEFAULT_NORM), ::ODEProblem{Array{Float64,1},Tuple{Float64,Float64},false,Array{Float64,1},ODEFunction{false,RCall.var"#11#12"{RObject{ClosSxp}},UniformScaling{Bool},Nothing,Nothing,Nothing,Nothing,Nothing,Nothing,Nothing,Nothing,Nothing,Nothing,Nothing,Nothing},Base.Iterators.Pairs{Union{},Union{},Tuple{},NamedTuple{,Tuple{}}},DiffEqBase.StandardODEProblem}, ::OrdinaryDiffEq.ODEIntegrator{CompositeAlgorithm{Tuple{Tsit5,Rosenbrock23{0,false,DefaultLinSolve,DataType}},OrdinaryDiffEq.AutoSwitchCache{Tsit5,Rosenbrock23{0,false,DefaultLinSolve,DataType},Rational{Int64},Int64}},false,Array{Float64,1},Nothing,Float64,Array{Float64,1},Float64,Float64,Float64,Array{Array{Float64,1},1},OrdinaryDiffEq.ODEComposite 
```

**Comment by ChrisRackauckas** [comment]
```julia
julia> versioninfo()
Julia Version 1.5.0
Commit 96786e22cc (2020-08-01 23:44 UTC)
Platform Info:
  OS: Windows (x86_64-w64-mingw32)
  CPU: Intel(R) Core(TM) i7-8550U CPU @ 1.80GHz
  WORD_SIZE: 64
  LIBM: libopenlibm
  LLVM: libLLVM-9.0.1 (ORCJIT, skylake)
```

is what it downloads. It looks like it doesn't get the patches? I don't know if that's the problem. But if I open up that binary and things directly in it then it seems fine:

```julia
julia> using DifferentialEquations

julia> f(u,p,t) = 1.01*u
f (generic function with 1 method)

julia> u0 = 1/2
0.5

julia> tspan = (0.0,1.0)
(0.0, 1.0)

julia> prob = ODEProblem(f,u0,tspan)
ODEProblem with uType Float64 and tType Float64. In-place: false
timespan: (0.0, 1.0)
u0: 0.5

julia> sol = solve(prob, Tsit5(), reltol=1e-8, abstol=1e-8)
retcode: Success
Interpolation: specialized 4th order "free" interpolation
t: 17-element Array{Float64,1}:
 0.0
 0.012407826196308189
 0.042501278333560696
 0.0817804940926822
 0.12887384498704435
 0.18409796286152927
 0.24627457447456758
 0.31479297816557983
 0.3885963515160237
 0.4668617724420117
 0.5487161305960653
 0.6334346972152323
 0.7203630000154827
 0.808957991167541
 0.8987655040395068
 0.9894161889652783
 1.0
u: 17-element Array{Float64,1}:
 0.5
 0.5063053789114713
 0.5219304750950854
 0.5430527156531716
 0.5695067765051011
 0.6021743618001767
 0.6412025634645868
 0.6871475244605603
 0.7403258398933249
 0.8012223416244076
 0.8702768595124217
 0.9480214633315531
 1.0350186537489083
 1.1319031171663114
 1.2393734610662304
 1.3582039071131542
 1.3728005076225747
```

**Comment by Non-Contradiction** [comment]
The issue is that the julia dlls cannot be located by R. 
In the `julia_setup()` function, this issue is mitigated by setting the current folder of R to the julia binary folder (and the `julia_setup` function resets the original current folder after the function). 
But the issue still exists when the julia binary folder is neither the current folder nor on the PATH.

**Comment by Non-Contradiction** [comment]
Here are the relevant lines in the julia_setup function:
https://github.com/Non-Contradiction/JuliaCall/blob/3bc4aa48a8f434353924303d269a76f2bcad332b/R/zzz.R#L92-L104

I'm still investigating on whether there is some easy way for R to find the julia dlls. Maybe I can `dyn.load` all the relevant julia dlls into R, but it doesn't seem a good way to do it...

**Comment by ChrisRackauckas** [comment]
The test case I'm running is https://github.com/SciML/diffeqr/pull/31. It's almost there 😄 .

Pinging @christopher-dG if he has an idea for that.

**Comment by Non-Contradiction** [comment]
For reference purposes, the Windows dll issue should be fixed by #150.
